suite/tests/client-interface/drbbdup-drwrap-test.appdll.c

```diff
@@ -56,7 +48,33 @@ run_tests(void)
 #else
     __asm__ __volatile__("nop; nop; nop; nop");
 #endif
+}
+
+int EXPORT
+wrapme(int x)
+{
+    print("%s: arg %d\n", __FUNCTION__, x);
+    if (x % 2 == 0) {
+        /* Switch in the middle of a wrapped function. */
+        switch_modes();
+    }
+    return x;
+}
+
+void
+run_tests(void)
+{
+    /* First, a regular pre-and-post-wrapped instance. */
+    print("first wrapme returned %d\n", wrapme(1));
+    /* Now we'll have a pre but not post b/c we'll swap in the middle. */
     print("second wrapme returned %d\n", wrapme(2));
+    /* Now we swap back and will have no pre or post. */
+    print("third wrapme returned %d\n", wrapme(2));
+    /* If we did not clean up drwrap state, we would see a pre and *two* posts as drwrap
+     * tries to catch up to the interrupted state.  With the proper cleanup, we have
```

Comment:
Do you mean there would be two posts triggered by the fourth wrapme below? Trying to understand how drwrap would remember to "catch up". (Didn't investigate too much on this, so  maybe I'm  missing something obvious).

Reply:
On a post-call point, drwrap always checks the stack value to see whether a call was skipped because of a tail-call or a longjmp or other actions.  That is how it "catches up": it invokes the post-call handlers for post-call points that implicitly happened on the skip.